consensus/parlia/parlia.go

```diff
@@ -1685,38 +1685,45 @@ func (p *Parlia) GetJustifiedHeader(chain consensus.ChainHeaderReader, header *t
 		return nil
 	}
 
-	// There won't any attestation in first two blocks, return root.
-	if header.Number.Uint64() < 2 {
-		return chain.GetHeaderByNumber(0)
-	}
-
 	snap, err := p.snapshot(chain, header.Number.Uint64(), header.Hash(), nil)
 	if err != nil {
 		log.Error("Unexpected error when getting snapshot",
 			"error", err, "blockNumber", header.Number.Uint64(), "blockHash", header.Hash())
 		return nil
 	}
 
-	// If there is no vote justified block, then return root.
+	// If there is no vote justified block, then return root or naturally justified block.
 	if snap.Attestation == nil {
-		return chain.GetHeaderByNumber(0)
+		if header.Number.Uint64() <= naturallyJustifiedDist {
+			return chain.GetHeaderByNumber(0)
+		}
+		// Return naturally justified block.
+		return FindAncientHeader(header, naturallyJustifiedDist, chain, nil)
 	}
-	// Return naturally justified block.
+
+	// If the latest vote justified block is too far, return naturally justified block.
 	if snap.Number-snap.Attestation.TargetNumber > naturallyJustifiedDist {
 		return FindAncientHeader(header, naturallyJustifiedDist, chain, nil)
 	}
 	//Return latest vote justified block.
 	return chain.GetHeaderByHash(snap.Attestation.TargetHash)
 }
 
-// GetFinalizedHeader returns highest finalized block header before the specific block,
-// the attestation within the specific block will be taken into account.
-func (p *Parlia) GetFinalizedHeader(chain consensus.ChainHeaderReader, header *types.Header) *types.Header {
+// GetFinalizedHeader returns highest finalized block header before the specific block.
+// It will first to find vote finalized block within the specific backward blocks, the maximum backward blocks is 21.
+// If the vote finalized block not found, return its previous backward block.
+func (p *Parlia) GetFinalizedHeader(chain consensus.ChainHeaderReader, header *types.Header, backward uint64) *types.Header {
 	if !chain.Config().IsLynn(header.Number) {
 		return chain.GetHeaderByNumber(0)
 	}
-	if chain == nil || header == nil || header.Number.Uint64() < 2 {
-		return chain.GetHeaderByNumber(0)
+	if chain == nil || header == nil {
+		return nil
+	}
+	if backward > types.NaturallyFinalizedDist {
+		backward = types.NaturallyFinalizedDist
+	}
+	if header.Number.Uint64() < backward {
```

Comment:
I am confused that `backward` can be compared with both `header.Number` and `NaturallyFinalizedDist`, I think they are two different concept.

Reply:
Compare with the header.Number is only to ensure we backward to genesis block and then we should stop.